### PR TITLE
Bulk Read

### PR DESCRIPTION
**Pull request by bsieber-mozilla**
## Proposed changes

Bulk read based on:
https://docs.google.com/document/d/1LZ0H3j6j7pduyIqYevugw8C_0A2iCMPJmxpkdq5hqiQ/edit#heading=h.e86nyocuyup7

## Types of changes

What types of changes does your code introduce?
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update (if none of the other choices apply)

## Checklist

- [x] I have read the [guides](https://github.com/mozilla-it/ctms-api/tree/main/guides)
- [x] I have followed the [Mozilla Lean Data Policies](https://www.mozilla.org/en-US/about/policy/lean-data/)
- [x] Lint and tests pass both locally and on the cicd with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

## Further comments

_Extracted from the gdocs_

The response will be JSON, like this pseudo-response:

```json
{
  "next": "https://ctms.example.com/updates?start=2021-03-31T20:30:00+00:00&end=2021-03-31T20:45:00+00:00&after=MjAyMS0wMy0zMVQyMDozMToyOC40MzE0MDIrMD&limit=10",
  "start": "2021-03-31T20:30:00+00:00",
  "end": "2021-03-31T20:45:00+00:00",
  "limit": 10,
  "items": [
    {
      ... // CTMS contact JSON
    }, ... // Additional CTMS contact JSON
  ]
}
```

The JSON items:
next - the URL to fetch the next set of records from the paginated set, or  null for the last page in the set. When present, the URL includes a new item, after, that helps determine the next set of items.
start - The start time for this paginated set, in UTC
end - The end time for this paginated set, which may have been picked or adjusted by the server
limit - The maximum number of contacts in items
items - a list of the full contacts. Each contact has the same representation as you'd get from the GET /ctms/ API. The contacts are ordered by update time (earliest first), and then by email ID (email.email_id).

### CODE REVIEW

**Comment by bsieber-mozilla** [comment]
Screen Shot 2021-04-01 at 2 26 10 PM

File: ctms/crud.py
Comment (bsieber-mozilla) [inline ctms/crud.py:89]:
```suggestion
            Email.email_id > after_email_id,
```

**Comment by bsieber-mozilla** [comment]
### What's Next:

Testing!

## Reviewers:
@jwhitlock / @imbstack, I'm not doing a TDD approach here, but I'll be adding tests shortly. I think the functionality on the document and discussed should be featured here. I did some brief local testing _(only adding 3 contacts, but using limits of 1)_ with the make-file staging of the web and db services.

**Comment by imbstack** [comment]
Awesome! Probably for the manual testing of many rows, we can land my PR and then use it to populate the db for your stuff. Thanks for hopping on this so quickly

File: ctms/app.py
Comment (bsieber-mozilla) [inline ctms/app.py:406]:
```suggestion
    limit: Optional[int] = 10,
```

File: ctms/app.py
Comment (bsieber-mozilla) [inline ctms/app.py:421]:
```suggestion

```

File: ctms/app.py
Comment (bsieber-mozilla) [inline ctms/app.py:168]:
```suggestion
```

File: ctms/app.py
Comment (bsieber-mozilla) [inline ctms/app.py:196]:
```suggestion
```

File: ctms/crud.py
Comment (bsieber-mozilla) [inline ctms/crud.py:85]:
```suggestion
```

File: ctms/schemas/contact.py
Comment (bsieber-mozilla) [inline ctms/schemas/contact.py:144]:
```suggestion
    next: Optional[Union[AnyUrl, str]] = None
    items: List[Optional[CTMSResponse]] 
```

File: tests/unit/test_crud.py
Comment (bsieber-mozilla) [inline tests/unit/test_crud.py:2]: https://www.youtube.com/watch?v=4BQLE_RrTSU

**Comment by bsieber-mozilla** [comment]
(Force pushed to re-write history to be on top of Bstack's commits instead of them showing up in this PR)

File: pyproject.toml
Comment (bsieber-mozilla) [inline pyproject.toml:33]: for date-parsing without getting too complicated

File: tests/unit/test_crud.py
Comment (bsieber-mozilla) [inline tests/unit/test_crud.py:81]: Not sure if I used the StatementWatcher here correctly. Tried to follow the patterns y'all had in place for pytest fixtures and dealing with db queries.

Comment (jwhitlock) [reply tests/unit/test_crud.py:81]: yep, this is how it is supposed to be used, as a context manager when you want to confirm there weren't unexpected SQL queries

Comment (jwhitlock) [reply tests/unit/test_crud.py:81]: I am using ``watcher`` instead of ``sw`` as the variable name, because pylint is complaining about short names. For the lint check, we're telling pylint to check unit tests, but my editor is running it on every save.

**Comment by bsieber-mozilla** [comment]
Force pushed and dealt with some merge conflicts in crud.py and test_api.

File: ctms/crud.py
Comment (bsieber-mozilla) [inline ctms/crud.py]: This was accidentally applied, during the rebase conflict, but removed in the following commit.

Comment (bsieber-mozilla) [reply ctms/crud.py]: Looks like this didn't get properly "pre-committed" as it went through the rebase path. Going to force a commit to align this to standards

**Comment by bsieber-mozilla** [comment]
I can open this up in a new PR if it's cleaner... 
(Or if next time you'd like that as a reviewer, give this a rocket 🚀 and I'll opt for that)

File: ctms/schemas/__init__.py
Comment (bsieber-mozilla) [inline ctms/schemas/__init__.py:9]: I saw @jwhitlock add to the init for the PATCH, didn't realize this __init__ was in use, but added the two responses here. A CTMSBulkResponse has in it a List[CTMSOptionalResponse] which allows for ContactSchema or CTMSSingleResponse schemas to not have all parts of a contact, including amo or other by-table data

File: tests/unit/test_crud.py
Comment (bsieber-mozilla) [reply tests/unit/test_crud.py:2]: _(click only if you're ready for the time of your life)_

**Comment by bsieber-mozilla** [comment]
### What Remains

**Review!**
(+any modifications based on review)

File: ctms/schemas/contact.py
Comment (jwhitlock) [inline ctms/schemas/contact.py:169]: Should this be ``List[CTMSOptionalResponse]``? I'm not sure what the ``Optional`` is for.

File: ctms/schemas/contact.py
Comment (jwhitlock) [inline ctms/schemas/contact.py:131]: I think the docstring needs updating, looks like it was copied.

I'm OK to use this, but we'll need to check with Cinchy if it is OK if the responses are different from ``/ctms`` and ``/ctms/{email_id}``. Here's what a record with missing group data looks like:

```json
{
  "items": [
    {
      "amo": null,
      "email": {
        "primary_email": "jwhitlock@mozilla.com",
        "basket_token": "8b359504-d1b4-4691-9175-cfee3059b171",
        "double_opt_in": false,
        "sfdc_id": null,
        "first_name": null,
        "last_name": null,
        "mailing_country": "us",
        "email_format": "T",
        "email_lang": "es",
        "has_opted_out_of_email": false,
        "unsubscribe_reason": null,
        "email_id": "e1d35779-9f14-4553-b2aa-85f9629f68bb",
        "create_timestamp": "2021-04-07T16:56:36.579296+00:00",
        "update_timestamp": "2021-04-07T16:56:36.579296+00:00"
      },
      "fxa": null,
      "mofo": null,
      "newsletters": [],
      "vpn_waitlist": null
    }
]
}
```

With the other types, the ``"fxa": null,`` and similar items would instead be the default values.

File: tests/unit/test_api.py
Comment (jwhitlock) [inline tests/unit/test_api.py:367]: Thanks for adding this!

File: ctms/app.py
Comment (jwhitlock) [inline ctms/app.py:237]: Nit: This branch is untested.

File: ctms/app.py
Comment (jwhitlock) [inline ctms/app.py:226]: This works, and there is a next URL until no results are returned. I'm thinking if ``len(results) < limit`` would work as well - results for the last set, but no next URL. However, the machine calling this API probably doesn't care.

**Review by jwhitlock — APPROVED** [review]
Thanks @bsieber-mozilla, the code and the tests look good, and it worked locally for me in the interactive docs.

We should notify Cinchy of the changes from the original spec:

* ``/bulk`` vs ``/updates``
* The difference in contact JSON between ``/ctms`` and ``/bulk``
* The addition of the ``after`` parameter to the body

I think all would be acceptable, and could be adjusted if they are a big problem. I think it is also easier to talk about it once it ships, so I think the order is merge and ship to stage then notify.

File: ctms/schemas/contact.py
Comment (bsieber-mozilla) [reply ctms/schemas/contact.py:169]: The intent of this optional was for a blank list; as a List[CTMSOptionalResponse] I think still had a requirement on their being some inner value. Can modify to test and confirm/deny

File: ctms/schemas/contact.py
Comment (bsieber-mozilla) [reply ctms/schemas/contact.py:131]: Ah, I wasn't 100% sure what should happen in this case. I think maybe the removal of the optional class could align

File: ctms/app.py
Comment (bsieber-mozilla) [inline ctms/app.py]: Removed!

File: ctms/app.py
Comment (bsieber-mozilla) [inline ctms/app.py:227]: added a page_length check to not create a "next" if on last-page

File: ctms/app.py
Comment (bsieber-mozilla) [inline ctms/app.py:454]: changed to map to gdocs provided to cinchy

File: ctms/crud.py
Comment (bsieber-mozilla) [inline ctms/crud.py:26]: move into relative `.schemas` from `ctms.schemas`

File: ctms/schemas/contact.py
Comment (bsieber-mozilla) [inline ctms/schemas/contact.py:131]: I think I can remove this object... in the next commit

File: ctms/schemas/contact.py
Comment (bsieber-mozilla) [inline ctms/schemas/contact.py]: Swapping types here to `ContactSchema`

File: tests/unit/test_api.py
Comment (bsieber-mozilla) [inline tests/unit/test_api.py:57]: added a no results/no next path

File: tests/unit/test_api.py
Comment (bsieber-mozilla) [reply tests/unit/test_api.py:394]: Modified this from `/bulk` to `/updates`

**Review by bsieber-mozilla — COMMENTED** [review]
Went through previous review comments and addressed some as changes!

File: ctms/app.py
Comment (bsieber-mozilla) [reply ctms/app.py]: Modified the code for this

**Comment by bsieber-mozilla** [comment]
@jwhitlock, thanks for the review and slack conversation around this. Tomorrow whenever you've got a moment to do a final review, it would be most appreciated! I'm thinking this PR should now solve most/all of the concerns. 

_This being my first PR into the repo, a once-over on the recent commit or anything I might've missed elsewhere would be the last step I feel before this moves over into `main`_

**Review by jwhitlock — APPROVED** [review]
Thanks for the updates @bsieber-mozilla, and for the fast work.

I tried a few things locally, with a database with plenty of records:
* Made a query that returned no results
* Made a query that returned some results with pagination
* Made a query that returned 1000 results (about 1.2 seconds locally)
* Changed parameters so they didn't match `next` (`next`'s start time is used)

Everything worked as expected.  🚢 SHIP IT

File: ctms/crud.py
Comment (bsieber-mozilla) [reply ctms/crud.py]: Redacted

**Comment by bsieber-mozilla** [comment]
Made that last change, green build; shipping it! :shipit:

**Pull request merged**
